src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/Inflater.cs

```diff
@@ -84,7 +86,14 @@ public unsafe int InflateVerified(byte* bufPtr, int length)
                 int bytesRead;
                 if (ReadInflateOutput(bufPtr, length, ZLibNative.FlushCode.NoFlush, out bytesRead) == ZLibNative.ErrorCode.StreamEnd)
                 {
-                    _finished = true;
+                    if (!NeedsInput() && IsGzipStream() && _inputBufferHandle.IsAllocated)
+                    {
+                        _finished = ResetStreamForLeftoverInput();
+                    }
+                    else
+                    {
+                        _finished = true;
+                    }
```

Comment:
If you wanted, seems like this could now be combined to be something like:
```C#
_finished = NeedsInput() || !IsGzipStream() || !_inputBufferHandle.IsAllocated || ResetStreamForLeftoverInput();
```